File: backend/src/lib/response.ts
import type { APIGatewayProxyResult } from "aws-lambda";
import type { ApiResponse } from "@notohub/shared";
import { config } from "./config.js";

const corsHeaders = {
  "Access-Control-Allow-Origin": config.allowedOrigin,
  "Access-Control-Allow-Headers": "Content-Type,Authorization",
  "Access-Control-Allow-Methods": "GET,POST,OPTIONS",
};

export function ok<T>(data: T, statusCode = 200): APIGatewayProxyResult {
  const body: ApiResponse<T> = { success: true, data };
  return {
    statusCode,
    headers: { "Content-Type": "application/json", ...corsHeaders },
    body: JSON.stringify(body),
  };
}

export function notFound(message = "Not found"): APIGatewayProxyResult {
  const body: ApiResponse<never> = { success: false, error: message };
  return {
    statusCode: 404,
    headers: { "Content-Type": "application/json", ...corsHeaders },
    body: JSON.stringify(body),
  };
}

export function serverError(err: unknown): APIGatewayProxyResult {
  const message =
    err instanceof Error ? err.message : "Internal server error";
  console.error("[ERROR]", err);
  const body: ApiResponse<never> = { success: false, error: message };
  return {
    statusCode: 500,
    headers: { "Content-Type": "application/json", ...corsHeaders },
    body: JSON.stringify(body),
  };
}

export function noContent(): APIGatewayProxyResult {
  return { statusCode: 204, headers: corsHeaders, body: "" };
}

File: backend/lib/response.py
import json
from decimal import Decimal
from typing import Any

from lib.config import config


def _cors_headers() -> dict[str, str]:
    return {
        "Access-Control-Allow-Origin": config.allowed_origin,
        "Access-Control-Allow-Headers": "Content-Type,Authorization",
        "Access-Control-Allow-Methods": "GET,POST,OPTIONS",
    }


class _DecimalEncoder(json.JSONEncoder):
    """boto3 returns DynamoDB numbers as Decimal; coerce them to float."""

    def default(self, o: Any) -> Any:
        if isinstance(o, Decimal):
            return float(o)
        return super().default(o)


def _json(body: Any) -> str:
    return json.dumps(body, cls=_DecimalEncoder)


def ok(data: Any, status_code: int = 200) -> dict:
    return {
        "statusCode": status_code,
        "headers": {"Content-Type": "application/json", **_cors_headers()},
        "body": _json({"success": True, "data": data}),
    }


def not_found(message: str = "Not found") -> dict:
    return {
        "statusCode": 404,
        "headers": {"Content-Type": "application/json", **_cors_headers()},
        "body": _json({"success": False, "error": message}),
    }


def server_error(exc: Exception) -> dict:
    print(f"[ERROR] {exc}", flush=True)
    return {
        "statusCode": 500,
        "headers": {"Content-Type": "application/json", **_cors_headers()},
        "body": _json({"success": False, "error": str(exc)}),
    }


def no_content() -> dict:
    return {"statusCode": 204, "headers": _cors_headers(), "body": ""}


def unauthorized(message: str = "Unauthorized") -> dict:
    return {
        "statusCode": 401,
        "headers": {"Content-Type": "application/json", **_cors_headers()},
        "body": _json({"success": False, "error": message}),
    }


def bad_request(message: str) -> dict:
    return {
        "statusCode": 400,
        "headers": {"Content-Type": "application/json", **_cors_headers()},
        "body": _json({"success": False, "error": message}),
    }


def method_not_allowed() -> dict:
    return {
        "statusCode": 405,
        "headers": {"Content-Type": "application/json", **_cors_headers()},
        "body": _json({"success": False, "error": "Method not allowed"}),
    }
